### PR TITLE
[BEAM-2264] Reuse GCP credentials in GCS calls.

### PR DESCRIPTION
**Pull request by udim**
I've preserved the @retry.with_exponential_backoff wrapper around credential refreshes out of an abundance of caution. The exception types are the same AFAICT from reading the code.

Commit notes:
- Uses apitools' thread-safe GCE credentials class instead of Beam's.
- Removes thread-local storage from gcsio, which was not used correctly
and when it was, surfaced possible race conditions in storage_v1_client
[BEAM-3990].
- get_service_credentials() now caches a single credentials object
forever.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by udim** [comment]
R: @lukecwik (since you have some state already, lmk if you don't have bandwidth)

File: sdks/python/apache_beam/internal/gcp/auth.py [hunk 1]
```diff
@@ -57,77 +71,58 @@ def set_running_in_gce(worker_executing_project):
...
+
+class _Credentials(object):
+  _credentials_lock = threading.Lock()
+  _credentials_init = False
+  _credentials = None
+
+  @classmethod
+  def get_service_credentials(cls):
+    if cls._credentials_init:
+      return cls._credentials
```
Comment (aaltay) [inline sdks/python/apache_beam/internal/gcp/auth.py]: Is it possible for credentials to expire?

Comment (udim) [reply sdks/python/apache_beam/internal/gcp/auth.py]: yes; they auto-refresh

**Comment by udim** [comment]
run python postcommit

**Comment by robertwb** [comment]
run python postcommit

**Comment by robertwb** [comment]
Run Python 3.6 PostCommit

**Comment by robertwb** [comment]
Run Python 2 PostCommit
Run Python 3.5 PostCommit
Run Python 3.7 PostCommit

**Comment by udim** [comment]
run python 3.6 postcommit

**Comment by udim** [comment]
3.6 failure is https://issues.apache.org/jira/browse/BEAM-7814

**Comment by udim** [comment]
I still want to run a performance test before merging

**Comment by udim** [comment]
Run Python27 WordCountIT Performance Test

**Comment by udim** [comment]
Stats for performance test: https://apache-beam-testing.appspot.com/explore?dashboard=5691127080419328

**Comment by udim** [comment]
run seed job

**Comment by udim** [comment]
Run Python27 WordCountIT Performance Test

**Comment by udim** [comment]
No difference in wordcount (~1000s run time). Trying internal tests.

**Comment by udim** [comment]
Internal tests seem to have a constant amount of `refreshing access_token` messages (~200 over 2h). Going ahead with merge.

**Comment by datadidit** [comment]
Is this out? I don't see 2.15.0 in pypi. Also do you all have any input on when DataFlow updates it's Beam version. This seems like a big feature for anyone doing PubSub -> ReadAllFromText() with Python

**Comment by udim** [comment]
> Is this out? I don't see 2.15.0 in pypi. Also do you all have any input on when DataFlow updates it's Beam version. This seems like a big feature for anyone doing PubSub -> ReadAllFromText() with Python

2.15 is not yet released. (see: https://lists.apache.org/thread.html/371f94b123857c87b0c0652f5ab0f8780bb629b075a9d1a6c701b110@%3Cdev.beam.apache.org%3E)
I'm not sure what you mean by Dataflow's Beam version. Dataflow should start supporting the latest Beam version right after it's released.

**Comment by datadidit** [comment]
@udim thanks for getting back to me. I didn't realize they were in lockstep https://cloud.google.com/dataflow/docs/support/sdk-version-support-status#apache-beam-sdks-2x . Do you know if the same day beam is release DataFlow gets the update or is there usually a delay. Mainly curious

**Comment by aaltay** [comment]
@datadidit So far all Beam releases has been same day available on Dataflow.

**Comment by datadidit** [comment]
@aaltay thanks!

**Pull request merged**